autorotate in-memory jpegs

brian-armstrong-discord opened the pull request:
resolves #8172

This pull request changes ExifReader to operate on istreams rather than filenames. It then builds a streambuf on top of Mat so that imdecode may use ApplyExifOrientation like imread does.

brian-armstrong-discord commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -284,6 +295,51 @@ static void ApplyExifOrientation(const String& filename, Mat& img)
+    Size bufSize = buf.size();
+    if( bufSize.width == 1 && buf.type() == CV_8UC1 )
+    {
+       ByteStreamBuffer bsb( reinterpret_cast<char*>(buf.data), bufSize.height );
```
ahhh ok that makes sense 👍 i still want to check for `CV_8UC1` though, right?

brian-armstrong-discord commented:
The test that failed here doesn't seem related to this patch

`OpenCV: FFMPEG: tag 0x34363248/'H264' is not supported with codec id 28 and format 'mp4 / MP4 (MPEG-4 Part 14)'
OpenCV: FFMPEG: fallback to use tag 0x00000021/'!???'`

Could you help me find what I might have broke?

alalek commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -284,6 +306,50 @@ static void ApplyExifOrientation(const String& filename, Mat& img)
+    if( buf.isContinuous() )
+    {
+       ByteStreamBuffer bsb( reinterpret_cast<char*>(buf.data), buf.total() * buf.elemSize() );
+       std::istream stream( &bsb );
```
Looks like the current `ByteStreamBuffer` implementation is buggy.

Please use this instead (with added `#include <sstream>` and removed `ByteStreamBuffer`):
```
std::istringstream stream(std::string(reinterpret_cast<char*>(buf.data), buf.total() * buf.elemSize()), std::ios_base::in);
```

(it does extra `memcpy` operation, but it is not critical for I/O tasks)

brian-armstrong-discord replied:
i'm not sure this is what caused build to break - it passed with a nearly identical implementation

brian-armstrong-discord commented:
@alalek Is there any way to retrigger the builds? I'm curious if this build error is going to happen consistently. The builds all passed previously, I think I barely made any changes after that

brian-armstrong-discord commented:
@alalek Do you feel pretty confident that's what caused it to break? The build did pass on a previous commit, though with the squash the changes have been hidden. IIRC the only things I changed were that I added an additional bounds check for `INT_MIN` and added your suggestion to use `buf.total() * bufSize.elemSize()`. That doesn't seem like enough to cause video files? to fail - in fact, after checking the test in question, I'm not even sure imread/imdecode are being invoked.

alalek commented:
Stack trace for `ByteStreamBuffer` ctor call:
```
#0  0x00000000005a3ef3 in (anonymous namespace)::ByteStreamBuffer::ByteStreamBuffer(char*, std::size_t) (this=0x7ffffffecf50, base=0x1908bf0 "\377\330\377", <incomplete sequence \340>, length=38608) at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:66
#1  0x00000000005a5c97 in cv::ApplyExifOrientation(cv::Mat const&, cv::Mat&) (buf=..., img=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:339
#2  0x00000000005a73c1 in cv::imdecode(cv::_InputArray const&, int) (_buf=..., flags=3)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:733
#3  0x000000000058f720 in cv::MotionJpegCapture::retrieveFrame(int, cv::_OutputArray const&) (this=0x1908840, output_frame=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap_mjpeg_decoder.cpp:824
#4  0x0000000000582bbb in cv::VideoCapture::retrieve(cv::_OutputArray const&, int) (this=0x7ffffffed4d0, image=..., channel=0)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:640
#5  0x0000000000582d96 in cv::VideoCapture::read(cv::_OutputArray const&) (this=0x7ffffffed4d0, image=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:663
#6  0x0000000000582dfb in cv::VideoCapture::operator>>(cv::Mat&) (this=0x7ffffffed4d0, image=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:691
#7  0x00000000004fc681 in CV_PositioningTest::run(int) (this=0x7fffffffd6f0)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/test/test_video_pos.cpp:142
```

It is called from "MotionJpegCapture" in this test.
BTW, Probably it is better to use `IMREAD_IGNORE_ORIENTATION` flag in MJPEG decoder.

Crash backtrace (with [log messages](https://github.com/alalek/opencv/commit/f75528ec5410c043baf4dd29cec44f9ccba43a03)):
```
ByteStreamBuffer(0x18f2310, 28272)
ByteStreamBuffer::seekoff(14, 1, ...)
ByteStreamBuffer::seekoff(65, 1, ...)
ByteStreamBuffer::seekoff(65, 1, ...)
ByteStreamBuffer::seekoff(29, 1, ...)
ByteStreamBuffer::seekoff(179, 1, ...)
ByteStreamBuffer::seekoff(29, 1, ...)
ByteStreamBuffer::seekoff(179, 1, ...)
ByteStreamBuffer::seekoff(15, 1, ...)
ByteStreamBuffer::seekoff(10, 1, ...)
ByteStreamBuffer::seekoff(35366, 1, ...)

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff102d970 in __memcpy_sse2_unaligned () from /lib64/libc.so.6
(gdb) bt
#0  0x00007ffff102d970 in __memcpy_sse2_unaligned () at /lib64/libc.so.6
#1  0x00007ffff7b7833b in std::basic_streambuf<char, std::char_traits<char> >::xsgetn(char*, long) () at /lib64/libstdc++.so.6
#2  0x00007ffff7b51b9b in std::istream::read(char*, long) () at /lib64/libstdc++.so.6
#3  0x00000000005ca5c0 in cv::ExifReader::getExif() (this=0x7ffffffecf90)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/exif.cpp:134
#4  0x00000000005ca1ba in cv::ExifReader::parse() (this=0x7ffffffecf90) at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/exif.cpp:83
#5  0x00000000005a5cd4 in cv::ApplyExifOrientation(cv::Mat const&, cv::Mat&) (buf=..., img=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:342
#6  0x00000000005a73c1 in cv::imdecode(cv::_InputArray const&, int) (_buf=..., flags=3)
    at /home/alalek/projects/opencv/opencv_dev/modules/imgcodecs/src/loadsave.cpp:733
#7  0x000000000058f720 in cv::MotionJpegCapture::retrieveFrame(int, cv::_OutputArray const&) (this=0x1908840, output_frame=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap_mjpeg_decoder.cpp:824
#8  0x0000000000582bbb in cv::VideoCapture::retrieve(cv::_OutputArray const&, int) (this=0x7ffffffed4d0, image=..., channel=0)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:640
#9  0x0000000000582d96 in cv::VideoCapture::read(cv::_OutputArray const&) (this=0x7ffffffed4d0, image=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:663
#10 0x0000000000582dfb in cv::VideoCapture::operator>>(cv::Mat&) (this=0x7ffffffed4d0, image=...)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/src/cap.cpp:691
#11 0x00000000004fc681 in CV_PositioningTest::run(int) (this=0x7fffffffd6f0)
    at /home/alalek/projects/opencv/opencv_dev/modules/videoio/test/test_video_pos.cpp:14
```


The similar logs messages are on [buildbot](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64_no_opt/builds/10464/steps/test_videoio/logs/stdio) (search for "Failure" keyword).

brian-armstrong-discord commented:
Ah, that helps a lot.

I don't see stacktraces in the buildbot output. Maybe I don't have enough permissions? All I see is

```/build/precommit_linux64_no_opt/opencv/modules/ts/src/ts.cpp:540: Failure
Failed

	failure reason: Arithmetic exception
	test case #-1
	seed: ffffffffffffffff
-----------------------------------
	LOG:

File: /tmp/__opencv_temp.jNDKKz/__opencv_temp.YQfV7G.XVID.avi

File: /tmp/__opencv_temp.jNDKKz/__opencv_temp.UJvnn8.MPEG.avi

File: /tmp/__opencv_temp.jNDKKz/__opencv_temp.loKZRG.MJPG.avi
General failure:
	Arithmetic exception (-6)

-----------------------------------```

which didn't really tell me anything :)

alalek commented:
`ByteStreamBuffer` has very limited implementation. Looks like it handles `dir == std::ios_base::cur` case only. Unfortunately there are no reliable checks for unsupported parameters (just several `return -1` statements for some cases).
So I still prefer to avoid custom implementation of mapped byte stream and to use `std::istringstream` instead.

brian-armstrong-discord commented:
@alalek If you look at ExifReader.cpp, you can see that the only seeking it does (with `seekg`) is based on the current pointer, so in fact this is fine. If I changed ExifReader to inspect the return values from these `seekg` calls, would that work for you? It seems silly to me to allocate and copy a bunch of memory just for this purpose

brian-armstrong-discord commented:
@alalek Also, for what it's worth, `seekoff`'s base behavior is to return -1, so doing so here merely conforms to that. http://www.cplusplus.com/reference/streambuf/streambuf/seekoff/

brian-armstrong-discord commented:
@alalek I've pushed a new version which will check the result of `seekg`. I think this is in fact safer than the previous version which never checked the result of `seek`

alalek commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -50,10 +50,55 @@
+         {
+            return -1;
+         }
+         gbump( static_cast<int>(off) );
```
"int" type limitation can be dropped (with `INT_MAX`/`INT_MIN` checks above):
```
setg(eback(), gptr() + off, egptr());
```

brian-armstrong-discord replied:
it can't be dropped because the windows build complains about silently converting a long to an int

alalek replied:
I mean this:
```.diff
- gbump( static_cast<int>(off) );
+ setg(eback(), gptr() + off, egptr());
```
There is no conversion to "int".

brian-armstrong-discord replied:
👍

brian-armstrong-discord commented:
@alalek Thanks for all your help so far 👍 I know it's been kind of a struggle, heh

Let me know if there's anything else I can do to help this PR along

alalek commented:
:+1:

ndtreviv commented:
Forgive my idiocy, but which version did this ship in? I've got 3.3.1 and it's still not autorotating images based on exif rotation settings via imdecode...

caogang commented:
This bug is fixed after which version?
